### PR TITLE
Forward requests to correct coordinator based on transaction header

### PR DESCRIPTION
**Pull request by danielhlarkin**
# Scope & Purpose

Currently we do not support load balancing when using the modern JS transactions API. We support request forwarding on specific APIs for coordinator-specific resources, but currently do not forward based on the transaction header. This PR attempts to rectify this.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] Bug-Fix for a *released version* (subsequent PRs forthcoming)

#### Related Information

- [x] There is an internal planning ticket: https://arangodb.atlassian.net/jira/servicedesk/projects/OASIS/queues/custom/233/OASIS-221

### Testing & Verification

This change is already covered by existing tests, such as load_balancing

Jenkins: https://jenkins.arangodb.biz/job/arangodb-matrix-pr/9855/

### CODE REVIEW

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/9881/

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/9923/

File: tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js [hunk 1]
```diff
@@ -250,6 +254,69 @@ function TransactionsSuite () {
       }
     },
 
+    testUseTransaction: function () {
+      let trx;
+      try {
+        const obj = { collections: { read: cn } };
+        let result = sendRequest('POST', "/_api/transaction/begin", obj, true);
+        assertEqual(result.status, 201);
+        assertFalse(result.body.result.id === undefined);
```
Comment (jsteemann) [inline tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js:263]: Simplification for next time (note: also requires adding `assertNotUndefined` to the list of globals in the prelude)
```suggestion
        assertNotUndefined(result.body.result.id);
```

File: tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js [hunk 2]
```diff
@@ -250,6 +254,69 @@ function TransactionsSuite () {
...
+        trx = result.body.result.id;
+        
+        // use trx on different coord to run a query
+        const query = `FOR doc IN ${cn} RETURN doc`;
+        result = sendRequest('POST', "/_api/cursor", { query }, false);
+        assertFalse(result.body.error);
+        assertEqual(10, result.body.result.length);
```
Comment (jsteemann) [inline tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js:270]: This tests requesting a cursor from a different coordinator, so it invokes the redirect. I think in addition we can also verify in the response that something was redirected. The HTTP response should send a special header to indicate this, but I don't know its name from the top of my head.
We should also add a test that tries to request the cursor from the correct coordinator (i.e. no redirect), and then verify that the header wasn't set.

File: tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js [hunk 3]
```diff
@@ -250,6 +254,69 @@ function TransactionsSuite () {
...
+      try {
+        const obj = { collections: { read: cn } };
+        let result = sendRequest('POST', "/_api/transaction/begin", obj, true);
+        assertEqual(result.status, 201);
+        assertFalse(result.body.result.id === undefined);
+        trx = result.body.result.id;
+        
+        // use trx on different coord to run a query
+        const query = `FOR doc IN ${cn} RETURN doc`;
+        result = sendRequest('POST', "/_api/cursor", { query }, false);
```
Comment (jsteemann) [inline tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js:268]: The query fully consumes the cursor at once. This is fine, but we should also add a test with a cursor that is queried incrementally, e.g. one result row at a time. This should be possible by giving the query a `batchSize` upon creation, and then issuing follow-up requests to `/_api/cursor/<id>`. This should work on both the correct coordinator and the incorrect one (which should do the redirect).

File: tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js [hunk 4]
```diff
@@ -250,6 +254,69 @@ function TransactionsSuite () {
...
+            }
+          ]
+        };
+        let result = sendRequest('POST', "/_api/gharial", graphDef, true);
+        assertEqual(result.status, 202);
+
+        const obj = { collections: { read: "startVertices" } };
+        result = sendRequest('POST', "/_api/transaction/begin", obj, true);
+        assertEqual(result.status, 201);
+        assertFalse(result.body.result.id === undefined);
```
Comment (jsteemann) [inline tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js:299]:
```suggestion
        assertNotUndefined(result.body.result.id);
```

File: tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js [hunk 5]
```diff
@@ -250,6 +254,69 @@ function TransactionsSuite () {
...
+        trx = result.body.result.id;
+
+        const doc = { value: true };
+        result = sendRequest('POST', "/_api/gharial/myGraph/vertex/startVertices", doc, true);
+        assertEqual(result.status, 202);
+        assertFalse(result.body.vertex._key === undefined);
```
Comment (jsteemann) [inline tests/js/client/load-balancing/load-balancing-transactions-noauth-cluster.js:305]:
```suggestion
        assertNotUndefined(result.body.vertex._key);
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9929/

**Pull request merged**